File: fm-database/fmdb/plan/allergies.py
"""Allergy resolution — the difference between "none" and "never asked".

Python mirror of ``fm-database-web/src/lib/fmdb/allergies.ts``. Keep the two in
lockstep: the TS side serves the client app and the meal-photo check, this side
serves the assessment gate, the letter generators and recipe selection, and a
client must not be allergy-screened on one surface and not the other.

``Client.known_allergies`` was read by ten consumers as though a populated list
meant allergies and an empty one meant none. Measured on the live roster
2026-08-03: 1 record of 21 was non-empty, and its value was ``['None']``. So
every consumer was asserting a negative nobody had established — including two
safety checks (``author_gate._check_allergies``'s HARD block and
``recipe_select.is_safe``'s allergen join), both of which were passing
everything by construction.

The field already carries three states; the code was collapsing them into two:

    declared — real allergens. Absolute: never suggest, never serve.
    none     — the client was ASKED and said no. A real negative screen.
    unknown  — empty because the question was never answered. NOT none, and
               never to be rendered as one.

WHY THIS IS NOT MERGED INTO ``foods_to_avoid``: that field is prose written for
a human and holds "Onion, Garlic" (a Jain preference), "Brinjal, Rice, Wheat"
(a protocol phase) and "Brinjal (used to get itchy tongue as a kid)" (an oral
allergy) in the same shape. Severity is not mechanically recoverable from it.
An allergy is *never*; a coach's exclusion is *for now*.
"""

from __future__ import annotations

import re
from typing import Any, Iterable, Optional

# Values a human writes to mean "I have none". Matched WHOLE, after lowering
# and stripping punctuation — never as a substring, or "none of the nuts"
# would register as a negative screen.
_NONE_SENTINELS = {
    "none",
    "nil",
    "no",
    "na",
    "nka",
    "nkda",
    "no allergies",
    "none known",
    "no known allergies",
    "no known drug allergies",
    "not known",
    "nothing",
    "no allergy",
    "denies",
}

#: Written when a client affirms they have none. Any of ``_NONE_SENTINELS``
#: reads back the same way, so a hand-typed "nil" from the coach also works.
NO_KNOWN_ALLERGIES = "No known allergies"

_PUNCT = re.compile(r"[^a-z ]+")
_WS = re.compile(r"\s+")


def _is_none_sentinel(raw: str) -> bool:
    s = _WS.sub(" ", _PUNCT.sub(" ", raw.lower())).strip()
    # Both forms: punctuation becomes a space, so "N/A" normalises to "n a" and
    # only matches once the spaces are also collapsed. "no known allergies"
    # needs the spaced form. Checking both covers each without a second list.
    return s in _NONE_SENTINELS or s.replace(" ", "") in _NONE_SENTINELS


def resolve_allergies(client: Optional[dict]) -> tuple[str, list[str]]:
    """Return ``(status, items)`` where status is declared | none | unknown.

    Reads ``known_allergies`` and falls back to ``allergies``. Both names exist
    in this codebase — ``updateClientProfile`` takes ``allergies`` as its input
    key and writes whichever key the file already has — and while all 21 live
    records currently use ``known_allergies``, a record created down the other
    branch would be missed by a single-name read.

    ``items`` is empty for BOTH none and unknown, so callers must branch on
    ``status``; testing ``if items:`` is the bug this module replaces.
    """
    client = client or {}

    def _seq(v: Any) -> Iterable:
        return v if isinstance(v, (list, tuple)) else ()

    raw = list(_seq(client.get("known_allergies"))) + list(_seq(client.get("allergies")))
    entries = [str(x).strip() for x in raw if str(x).strip()]
    if not entries:
        return "unknown", []

    items = [e for e in entries if not _is_none_sentinel(e)]
    # Every entry was a sentinel → asked, answered no. If a real allergen sits
    # alongside a "none", the allergen wins: ["none", "penicillin"] is a
    # data-entry artefact, not a tie to break in favour of safety-off.
    if not items:
        return "none", []
    return "declared", items


def allergy_prompt_line(client: Optional[dict]) -> str:
    """One line of allergy context for an AI prompt.

    ``unknown`` says so out loud rather than being omitted — an absent line
    reads to a model exactly like a cleared one, which is the failure this
    module exists to stop. This replaces the old
    ``'Allergies: none known'`` fallback, which asserted a screen that had
    never happened.
    """
    status, items = resolve_allergies(client)
    if status == "declared":
        return (
            "Allergies (ABSOLUTE — never suggest or serve these, at any dose): "
            + ", ".join(items)
        )
    if status == "none":
        return "Allergies: the client was asked and reported none."
    return (
        "Allergies: NOT RECORDED — nobody has asked. Absence here is not "
        "clearance; do not treat this client as allergy-free."
    )

File: fm-database-web/src/lib/fmdb/allergies.ts
/**
 * Allergy resolution — the difference between "none" and "never asked".
 *
 * `Client.known_allergies` was read by ten consumers as though a populated
 * list meant allergies and an empty one meant none. Measured on the live
 * roster 2026-08-03: 1 record of 21 was non-empty, and its value was
 * `['None']`. So every consumer was reading "no allergies known" off a field
 * nobody had filled in — a negative asserted, never established. Two of those
 * consumers are safety checks (the assessment gate's HARD allergen block, the
 * recipe safety join), and both were silently passing everything.
 *
 * The fix is not a new field. It is noticing that the existing one already
 * carries three states, and that the code was collapsing them into two:
 *
 *   declared — real allergens. Absolute: never suggest, never serve.
 *   none     — the client was ASKED and said no. A real negative screen, and
 *              until now discarded: `meal-check` filtered the sentinel out,
 *              which threw away the only genuine "none" on the roster.
 *   unknown  — the field is empty because the question was never answered.
 *              NOT the same as none, and must never be rendered as one.
 *
 * Deriving this from `known_allergies` rather than adding a field is
 * deliberate: the field is already in `_APP_CLIENT_KEYS`, so nothing here
 * trips the staging allowlist trap and goes invisible on Fly.
 *
 * WHY THIS IS NOT MERGED INTO `foods_to_avoid`. That field is the de-facto
 * home of exclusions and it is prose written for a human — but it holds
 * "Onion, Garlic" (a Jain preference), "Brinjal, Rice, Wheat" (a protocol
 * phase) and "Brinjal (used to get itchy tongue as a kid)" (an oral allergy)
 * in the same string, in the same shape. Severity cannot be recovered from it
 * mechanically. An allergy is never; a coach's exclusion is for now — and the
 * consumers act on them differently, so they stay separate fields.
 */

/** Values a human writes to mean "I have none". Matched whole, after
 *  lowercasing and stripping punctuation — never as a substring, or
 *  "none of the nuts" would read as a negative screen. */
const NONE_SENTINELS = new Set([
  "none",
  "nil",
  "no",
  "na",
  "nka",
  "nkda",
  "no allergies",
  "none known",
  "no known allergies",
  "no known drug allergies",
  "not known",
  "nothing",
  "no allergy",
  "denies",
]);

/** The sentinel written when a client affirms they have none. Read back by
 *  `resolveAllergies` as `none`; anything in NONE_SENTINELS also works, so a
 *  hand-typed "nil" from the coach is understood too. */
export const NO_KNOWN_ALLERGIES = "No known allergies";

export type AllergyStatus = "declared" | "none" | "unknown";

export type ResolvedAllergies = {
  status: AllergyStatus;
  /** Real allergens only. Empty for `none` and `unknown` alike — always
   *  branch on `status`, never on `items.length`. */
  items: string[];
};

function isNoneSentinel(raw: string): boolean {
  const s = raw
    .toLowerCase()
    .replace(/[^a-z ]+/g, " ")
    .replace(/\s+/g, " ")
    .trim();
  // Both forms: punctuation becomes a space, so "N/A" normalises to "n a" and
  // only matches once the spaces are also collapsed. "no known allergies"
  // needs the spaced form. Checking both covers each without a second list.
  return NONE_SENTINELS.has(s) || NONE_SENTINELS.has(s.replace(/ /g, ""));
}

/**
 * Resolve a client record's allergy state.
 *
 * Reads `known_allergies` and falls back to `allergies`. Both names exist in
 * this codebase — `updateClientProfile` takes `allergies` as its input key and
 * writes whichever key the file already has — and while all 21 live records
 * currently use `known_allergies`, a record created down the other branch
 * would be missed by a single-name read.
 */
export function resolveAllergies(
  client:
    | { known_allergies?: unknown; allergies?: unknown }
    | null
    | undefined,
): ResolvedAllergies {
  const raw = [
    ...(Array.isArray(client?.known_allergies) ? client.known_allergies : []),
    ...(Array.isArray(client?.allergies) ? client.allergies : []),
  ];

  const entries = raw.map((x) => String(x ?? "").trim()).filter(Boolean);
  if (entries.length === 0) return { status: "unknown", items: [] };

  const items = entries.filter((e) => !isNoneSentinel(e));
  // Every entry was a sentinel → the client was asked and said no. If any real
  // allergen sits alongside a "none", the allergen wins: a list reading
  // ["none", "penicillin"] is a data-entry artefact, not a contradiction to
  // resolve in favour of safety-off.
  if (items.length === 0) return { status: "none", items: [] };
  return { status: "declared", items };
}

/**
 * One line of allergy context for an AI prompt.
 *
 * `unknown` says so out loud rather than being omitted. An absent line reads
 * to a model exactly like a cleared one, which is the failure this whole
 * module exists to stop.
 */
export function allergyPromptLine(client: Parameters<typeof resolveAllergies>[0]): string {
  const { status, items } = resolveAllergies(client);
  if (status === "declared") {
    return `Allergies (ABSOLUTE — never suggest or serve these, at any dose): ${items.join(", ")}`;
  }
  if (status === "none") {
    return "Allergies: the client was asked and reported none.";
  }
  return "Allergies: NOT RECORDED — nobody has asked. Absence here is not clearance; do not treat this client as allergy-free.";
}

/**
 * Coach-facing label for the empty case, so no surface prints "None reported"
 * over a question that was never put to the client.
 */
export function allergyEmptyLabel(status: AllergyStatus): string {
  return status === "none" ? "None reported." : "Not recorded — never asked.";
}
